core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java

```diff
@@ -364,12 +346,25 @@ private int updateShardStatuses(EnumSet<ShardStatusChangeType> changes, List<Sha
             return changes;
         }
 
-        private void calcAndSetShardBuffers(int activeShards, String reason) {
-            if (activeShards == 0) {
+        private void calcAndSetShardBuffers(String reason) {
+
+            // Count how many shards are now active:
+            int activeShardCount = 0;
+            for (Map.Entry<ShardId,Boolean> ent : shardWasActive.entrySet()) {
+                if (ent.getValue()) {
+                    activeShardCount++;
+                }
+            }
+
+            // TODO: we could be smarter here by taking into account how RAM the IndexWriter on each shard
+            // is actually using (using IW.ramBytesUsed), so that small indices (e.g. Marvel) would not
+            // get the same indexing buffer as large indices.  But it quickly gets tricky...
```

Comment:
> Indeed! This sounds like a fun thing to think about!

I'll open a followon issue for this ... I think often there are very lopsided shards in a single node, and evenly dividing the indexing buffer seems crazy.

It would be better if the whole system acted like IW acts with N threads using up heap: as soon as the sum of all heap used by these threads exceeds the global budget, we ask the largest one to write to disk.